iml-services/iml-device/src/main.rs

```diff
@@ -106,29 +111,60 @@ async fn main() -> Result<(), ImlDeviceError> {
 
     tokio::spawn(server);
 
-    let mut s = consume_data::<Device>("rust_agent_device_rx");
+    let mut s = consume_data::<(Device, Vec<Command>)>("rust_agent_device_rx");
+    let mut i = 0usize;
+
+    while let Some((f, output)) = s.try_next().await? {
+        let begin: DateTime<Local> = Local::now();
+        tracing::info!("Iteration {}: begin: {}", i, begin);
+
+        let (d, cs) = output;
+
+        let mut incoming_devices = incoming_devices_2.lock().await;
+        incoming_devices.insert(f.clone(), d.clone());
+
+        assert!(
+            match d {
+                Device::Root(_) => true,
+                _ => false,
+            },
+            "The top device has to be Root"
+        );
+
+        let all_devices = incoming_devices
+            .iter()
+            .map(|(x, y)| (x.clone(), y.clone()))
+            .collect();
 
-    while let Some((f, d)) = s.try_next().await? {
-        let mut cache = cache2.lock().await;
+        let middle1: DateTime<Local> = Local::now();
 
-        cache.insert(f.clone(), d.clone());
+        let updated_devices = update_virtual_devices(all_devices, &cs);
 
-        let device_to_insert = NewChromaCoreDevice {
-            fqdn: f.to_string(),
-            devices: serde_json::to_value(d).expect("Could not convert incoming Devices to JSON."),
-        };
+        let middle2: DateTime<Local> = Local::now();
 
-        let new_device = diesel::insert_into(table)
-            .values(device_to_insert)
-            .on_conflict(fqdn)
-            .do_update()
-            .set(devices.eq(excluded(devices)))
-            .get_result_async::<ChromaCoreDevice>(&pool)
-            .await
-            .expect("Error saving new device");
+        let updated_devices_2 = updated_devices.clone();
+
+        let updated_devices: HashMap<Fqdn, Device> = updated_devices.clone().into_iter().collect();
+        let mut resolved_devices = resolved_devices_2.lock().await;
+        std::mem::replace(&mut *resolved_devices, updated_devices);
```

Comment:
There must be a warning about unused result of `replace`.

Reply:
No, why?